frontend/src/app/utils/deliveryPreferences.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  TestResultDeliveryPreference,
+  TestResultDeliveryPreferences,
+} from "../patients/TestResultDeliveryPreference";
+
+const { SMS, EMAIL, ALL, NONE } = TestResultDeliveryPreferences;
+
+export const toggleDeliveryPreferenceSms = toggleDeliveryPreference(SMS);
+export const toggleDeliveryPreferenceEmail = toggleDeliveryPreference(EMAIL);
+
+export const getSelectedDeliveryPreferencesSms = getSelectedDeliveryPreference(
+  SMS
+);
+export const getSelectedDeliveryPreferencesEmail = getSelectedDeliveryPreference(
+  EMAIL
+);
+
+function toggleDeliveryPreference(field: TestResultDeliveryPreference) {
+  return function handleToggle(
+    preference: TestResultDeliveryPreference,
+    newPreference: TestResultDeliveryPreference
+  ): TestResultDeliveryPreference {
+    const deliveryPreferences = new Set<TestResultDeliveryPreference>();
+
+    if (preference === SMS || preference === ALL) {
+      deliveryPreferences.add(SMS);
+    }
+
+    if (preference === EMAIL || preference === ALL) {
+      deliveryPreferences.add(EMAIL);
+    }
+
+    if (newPreference === field) {
+      deliveryPreferences.add(field);
+    }
+
+    if (newPreference === NONE) {
+      deliveryPreferences.delete(field);
+    }
+
```

Comment:
Maybe? 🙂 Like I was grousing in the PR notes, a patient's test delivery preferences are becoming complex enough that representing the preference state as a string is way underpowered, but it would take a lot of work to dig us out of that hole.

I'll think about a nicer way of doing this here, and of course I'm always open to suggestions. At the very least, some inline comments would help here.

Reply:
Let me know if the changes/comments added in 3613648 improve the readability here